server/controllers/authController.ts

```diff
@@ -0,0 +1,208 @@
+
+      if (disableHashPassword) {
+        match = req.user.fields[airtablePasswordColumn] === password;
+      } else {
+        match = await bcrypt.compare(
+          password,
+          req.user.fields[airtablePasswordColumn],
+        );
+      }
+      if (!match) {
+        return next(new AuthorizationError('Incorrect username or password'));
+      }
+
+      let token: string;
+      try {
+        token = createToken(req.user);
+      } catch (err) {
+        return next(err);
+      }
+      sendToken(req, res, token);
+    },
+
+    async register(req, res, next) {
+      if (!isEmpty(req.user)) {
+        return res.json({ success: false, message: 'User exists' });
+      }
+      if (!req.body.password || !req.body.password.trim()) {
+        return next(new InputError('No password was specified'));
+      }
+
+      const fields = req.body.fields ? req.body.fields : {};
+      let hash: string;
+      try {
+        hash = await bcrypt.hash(req.body.password, saltRounds);
+      } catch (err) {
+        return next(err);
+      }
+
+      let newUser: Record<any>;
+      try {
+        const {
+          body: { error: err, ...user },
+        } = await request({
+          url: AirtableRoute.users({
+            airtableBaseId,
+            airtableUserTableName,
+          }),
+          method: 'POST',
+          body: {
+            fields: {
+              ...fields,
+              [airtableUsernameColumn]: req.body.username,
+              [airtablePasswordColumn]: `${
+                disableHashPassword ? req.body.password : hash
+              }`,
+            },
+          },
+          ...requestOptions,
+        });
+        if (err) {
+          throw err;
+        }
+        newUser = user;
+      } catch (err) {
+        return next(err);
+      }
+
+      let token: string;
+      try {
+        token = createToken(newUser);
+      } catch (err) {
+        return next(err);
+      }
+      sendToken(req, res, token);
+    },
+
+    async checkForExistingUser(req, _res, next) {
+      if (!req.body || !req.body.username) {
+        return next();
+      }
+      const username = req.body.username;
+      const queryUserUrl = AirtableRoute.users(
+        {
+          airtableBaseId,
+          airtableUserTableName,
+        },
+        {
+          filterByFormula: `${airtableUsernameColumn}="${username}"`,
+        },
+      );
+
+      try {
+        const {
+          body: { records, error },
+          statusCode,
+        } = await request({
+          url: queryUserUrl,
+          method: 'GET',
+          ...requestOptions,
+        });
+        if (error || statusCode !== 200) {
+          throw new Error(
+            `[Airtable error: ${error?.type || ''}] ${error?.message ||
+              'unknown'}`,
+          );
+        }
+        [req.user] = records;
+        next();
+      } catch (err) {
+        next(err);
+      }
+    },
+
+    verifyToken(req, _res, next) {
```

Comment:
The non-authentication routes seemed to error out with `Can't set headers after they are sent`. This is sometimes caused by an unnecessary call to `next()` or instead of returning a successful payload, the code calls `next()` in a non-middleware context.

For us, it seems that we need to break this function explicitly into the `token` null and not null case for the last line in verifyToken to work.

Suggested revision is

```
     if (token) {
        jwt.verify(token, publicKey, (err, _decoded) => {
          if (err) {
            return next(new AuthorizationError('Invalid token supplied'));
          }
          return next();
        });
      } else {
        return next(new InputError('No token supplied'));
      }

```